File: src/main/java/dev/stashy/extrasounds/sounds/Sounds.java
package dev.stashy.extrasounds.sounds;

import net.minecraft.client.sound.Sound;
import net.minecraft.client.sound.SoundEntry;
import net.minecraft.sound.SoundEvent;
import net.minecraft.util.Identifier;

import java.util.List;

public class Sounds extends SoundRegistry
{
    public static final SoundEvent CHAT = register("chat.message");
    public static final SoundEvent CHAT_MENTION = register("chat.mention");
    public static final SoundEvent HOTBAR_SCROLL = register("hotbar_scroll");
    public static final SoundEvent INVENTORY_OPEN = register("inventory.open");
    public static final SoundEvent INVENTORY_CLOSE = register("inventory.close");
    public static final SoundEvent INVENTORY_SCROLL = register("inventory.scroll");
    public static final SoundEvent ITEM_DROP = register("item.drop");
    public static final SoundEvent ITEM_PICK = register("item.pickup");
    public static final SoundEvent ITEM_PICK_ALL = register("item.pickup_all");
    public static final SoundEvent ITEM_CLONE = register("item.clone");
    public static final SoundEvent ITEM_DELETE = register("item.delete");
    public static final SoundEvent ITEM_DRAG = register("item.drag");
    public static final SoundEvent EFFECT_ADD_POSITIVE = register("effect.add.positive");
    public static final SoundEvent EFFECT_ADD_NEGATIVE = register("effect.add.negative");
    public static final SoundEvent EFFECT_REMOVE_POSITIVE = register("effect.remove.positive");
    public static final SoundEvent EFFECT_REMOVE_NEGATIVE = register("effect.remove.negative");
    public static final SoundEvent KEYBOARD_TYPE = register("keyboard.type");

    public class Actions
    {
        public static final SoundEvent BOW_PULL = register("item.action.bow");
    }

    public static SoundEntry aliased(SoundEvent e)
    {
        return single(e.getId(), 1f, 1f, Sound.RegistrationType.SOUND_EVENT);
    }

    public static SoundEntry event(Identifier id)
    {
        return single(id, 1f, 1.7f, Sound.RegistrationType.SOUND_EVENT);
    }

    public static SoundEntry single(Identifier id, float volume, float pitch, Sound.RegistrationType type)
    {
        return new SoundEntry(List.of(
                new Sound(id.toString(), volume, pitch, 1, type, false, false, 16)
        ), false, null);
    }
}

File: src/main/kotlin/dev/stashy/extrasounds/sounds/Sounds.kt
package dev.stashy.extrasounds.sounds

import net.minecraft.client.sound.Sound
import net.minecraft.client.sound.Sound.RegistrationType
import net.minecraft.client.sound.SoundEntry
import net.minecraft.sound.SoundEvent
import net.minecraft.util.Identifier
import net.minecraft.util.math.floatprovider.ConstantFloatProvider


object Sounds : SoundRegistry {
    val CHAT: SoundEvent = register("chat.message")
    val CHAT_MENTION: SoundEvent = register("chat.mention")
    val HOTBAR_SCROLL: SoundEvent = register("hotbar_scroll")
    val INVENTORY_OPEN: SoundEvent = register("inventory.open")
    val INVENTORY_CLOSE: SoundEvent = register("inventory.close")
    val INVENTORY_SCROLL: SoundEvent = register("inventory.scroll")
    val ITEM_DROP: SoundEvent = register("item.drop")
    val ITEM_PICK: SoundEvent = register("item.pickup")
    val ITEM_PICK_ALL: SoundEvent = register("item.pickup_all")
    val ITEM_CLONE: SoundEvent = register("item.clone")
    val ITEM_DELETE: SoundEvent = register("item.delete")
    val ITEM_DRAG: SoundEvent = register("item.drag")
    val EFFECT_ADD_POSITIVE: SoundEvent = register("effect.add.positive")
    val EFFECT_ADD_NEGATIVE: SoundEvent = register("effect.add.negative")
    val EFFECT_REMOVE_POSITIVE: SoundEvent = register("effect.remove.positive")
    val EFFECT_REMOVE_NEGATIVE: SoundEvent = register("effect.remove.negative")
    val KEYBOARD_TYPE: SoundEvent = register("keyboard.type")

    object Actions {
        val BOW_PULL: SoundEvent = register("action.bow")
        val REPEATER_ADD: SoundEvent = register("action.repeater.add")
        val REPEATER_RESET: SoundEvent = register("action.repeater.reset")
    }

    fun aliased(e: SoundEvent): SoundEntry {
        return aliased(e, 1f)
    }

    fun aliased(e: SoundEvent, volume: Float): SoundEntry {
        return single(e.id, volume, 1f, RegistrationType.SOUND_EVENT)
    }

    fun event(id: Identifier): SoundEntry {
        return event(id, 0.6f)
    }

    fun event(id: Identifier, volume: Float): SoundEntry {
        return single(id, volume, 1.7f, RegistrationType.SOUND_EVENT)
    }

    fun single(id: Identifier, volume: Float, pitch: Float, type: RegistrationType?): SoundEntry {
        return SoundEntry(
            listOf(
                Sound(
                    id.toString(), ConstantFloatProvider.create(volume), ConstantFloatProvider.create(pitch), 1,
                    type, false, false, 16
                )
            ), false, null
        )
    }
}